### DISCUSSION TITLE
How to use pyodide npm package with vite

### DISCUSSION THREAD

**Discussion by rabbit19981023** [pyodide/pyodide#3673] [Q&A] [answered]
I installed pyodide via npm, and use it with vite, but always got err: `http://localhost:8000/node_modules/.vite/deps/pyodide_py.tar net::ERR_ABORTED 404 (Not Found)`, and `GET http://localhost:8000/node_modules/.vite/deps/pyodide.asm.js net::ERR_ABORTED 504 (Outdated Optimize Dep)`.

However, I found both `pyodide_py.tar` & `pyodide.asm.js` in `node_modules/pyodide`.

I have no idea how to include `pyodide_py.tar` & `pyodide.asm.js` in `node_modules/.vite/deps` folder to make vite recognizes them?

`src/main.ts`:
```typescript
import { loadPyodide } from "pyodide";

const pyodide = await loadPyodide();
const micropip = await pyodide.loadPackage('micropip');

// operations...
```

`vite.config.ts`
```typescript
import { defineConfig } from 'vite'
import vue from '@vitejs/plugin-vue'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [vue()],
})
```

**Comment by hoodmane** [comment]
@pauleveritt can you help?

**Comment by pauleveritt** [comment, marked as answer]
Yep, in fact, my talk Friday was about figuring out an npm install, Vite, Vitest, build vs. dev vs. test modes.

I wound up writing a tiny Vite plugin to automate copying into the bundle. I can share it, or share the whole sample project. One place I failed: I couldn't recreate the "re-use a Pyodide interpreter across all tests" like I previously had working.

**Reply by pauleveritt** [reply]
Here's the repo: https://github.com/pauleveritt/pwc-pyodide

**Reply by rabbit19981023** [reply]
Thank you for your reply!

It seems like there are still many works to do while developing with pyodide, I would write some tiny scripts like what your sample project did for the sake of simplicity and fast development.

Again, thank you a lot!

**Comment by siraisisatoru** [comment]
Hello.
I checked out the solution by @pauleveritt and it was working for Pyodide@v0.22.1 but not Pyodide@v0.25.0. With the help of StackOverflow and tones of GPT conversation, I managed to use Pyodide@v0.25.0 in the worker in Vite. You can take a look at the live [demo](https://siraisinotes-demo.web.app/test) or the GitHub repo [react-markdown-template](https://github.com/siraisisatoru/react-markdown-template). 

The goal of the template was to read markdown files, render code blocks, execute Python code and return the execution output to the main thread. To achieve this, react-markdown handles the file reading and I can process the code chunk by filtering the component. I initiated the worker in the markdown-render stage and passed the reference to the code block renderer to run Pyodide. The channel was established using Comlink. 

My template eliminates the issue of `importscripts` in workers by directly importing Pyodide. The template also tackles several warnings when importing the Pyodide in worker by installing `isomorphic-fetch` according to  [this StackOverflow answer](https://stackoverflow.com/questions/73095592/octokit-js-not-working-with-vite-module-externalized-and-cannot-be-accessed-in). 

Hopefully, you see some helpful from my example.

**Comment by castedo** [comment]
Readers interested in this topic might be interested in my conclusions at #5115 from struggling to get Pyodide to work with vite. In short, I decided Hot Module Replacement isn't worth it and I opted to use Mocha intead of `vitest`. I'm still using `vite build` + `vite preview`, for now, following instructions in the Pyodide docs. Although I'm tempted to just use Rollup to have one less tool dependency to fight with.
